In [11]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

import env

2. Given the following confusion matrix, evaluate (by hand) the model's performance.

|               | pred dog   | pred cat   |
|:------------  |-----------:|-----------:|
| actual dog    |         46 |         7  |
| actual cat    |         13 |         34 |


In [2]:
#establish baseline assuming a dog is a positive
#actual dog
46+7

53

In [3]:
#actual not dog
13+34

47

In [5]:
#baseline
53/(53+47)

0.53

#accuracy

TP + TN /
TP + TN + FP + FN


In [7]:
(46+34)/(46+34+13+7)

0.8

#precision

TP/TP +FP

In [8]:
46/(46+13)

0.7796610169491526

#recall

TP/TP+FN

In [9]:
46/(46+7)

0.8679245283018868

In the context of this problem, what is a false positive?



Predicting dog but it's actually a cat

In the context of this problem, what is a false negative?

Predicting cat but it's actually a dog

How would you describe this model?

I would say this model is very accurate based on accuracy, precision, and recall metrics.

3. You are working as a datascientist working for Codeup Cody Creator (C3 for short), a rubber-duck manufacturing plant.

Unfortunately, some of the rubber ducks that are produced will have defects. Your team has built several models that try to predict those defects, and the data from their predictions can be found here.

Use the predictions dataset and pandas to help answer the following questions:




In [14]:
df = pd.read_clipboard()
df.head()

actual     model1  model2     model3
0  No Defect  No Defect  Defect  No Defect
1  No Defect  No Defect  Defect     Defect
2  No Defect  No Defect  Defect  No Defect
3  No Defect     Defect  Defect     Defect
4  No Defect  No Defect  Defect  No Defect

An internal team wants to investigate the cause of the manufacturing defects. They tell you that they want to identify as many of the ducks that have a defect as possible.  

Which evaluation metric would be appropriate here?

Recall

Which model would be the best fit for this use case? model3

In [15]:
#baseline
(df.actual == 'Defect').mean()

0.08

In [19]:
#confusion matrix
counts1 = pd.crosstab(df.actual, df.model1)
counts2 = pd.crosstab(df.actual, df.model2)
counts3 = pd.crosstab(df.actual, df.model3)
counts1

model1     Defect  No Defect
actual                      
Defect          8          8
No Defect       2        182

In [21]:
tp1 = counts1.iloc[0,0] #rows, columns
fn1 = counts1.iloc[0,1]
fp1 = counts1.iloc[1,0]
tn1 = counts1.iloc[1,1]

In [22]:
tp1, fn1, fp1, tn1

(8, 8, 2, 182)

In [26]:
tp2 = counts2.iloc[0,0] #rows, columns
fn2 = counts2.iloc[0,1]
fp2 = counts2.iloc[1,0]
tn2 = counts2.iloc[1,1]

In [27]:
tp2, fn2, fp2, tn2

(9, 7, 81, 103)

In [30]:
tp3 = counts3.iloc[0,0] #rows, columns
fn3 = counts3.iloc[0,1]
fp3 = counts3.iloc[1,0]
tn3 = counts3.iloc[1,1]

In [31]:
tp3, fn3, fp3, tn3

(13, 3, 86, 98)

In [23]:
pos_actuals = df [df.actual == 'Defect']

In [28]:
#model1 recall
(pos_actuals.actual == pos_actuals.model1).mean()

0.5

In [29]:
#model2 recall
(pos_actuals.actual == pos_actuals.model2).mean()

0.5625

In [32]:
#model3 recall
(pos_actuals.actual == pos_actuals.model3).mean()

0.8125

model3 picked up 81.25% of the defects so it is the best fir in this case

Recently several stories in the local news have come out highlighting customers who received a rubber duck with a defect, and portraying C3 in a bad light. The PR team has decided to launch a program that gives customers with a defective duck a vacation to Hawaii. They need you to predict which ducks will have defects, but tell you the really don't want to accidentally give out a vacation package when the duck really doesn't have a defect.  

Which evaluation metric would be appropriate here?

Precision

Which model would be the best fit for this use case?

In [45]:
pos_predictions = df [df.model1 == 'Defect']

In [46]:
(pos_predictions.actual == pos_predictions.model1).mean()

0.8

In [47]:
pos_predictions = df [df.model2 == 'Defect']

In [48]:
(pos_predictions.actual == pos_predictions.model2).mean()

0.1

In [49]:
pos_predictions = df [df.model3 == 'Defect']

In [50]:
(pos_predictions.actual == pos_predictions.model3).mean()

0.13131313131313133

model1 is the best fit in this case based on it's precision

4. You are working as a data scientist for Gives You Paws ™, a subscription based service that shows you cute pictures of dogs or cats (or both for an additional fee).

At Gives You Paws, anyone can upload pictures of their cats or dogs. The photos are then put through a two step process. First an automated algorithm tags pictures as either a cat or a dog (Phase I). Next, the photos that have been initially identified are put through another round of review, possibly with some human oversight, before being presented to the users (Phase II).

Several models have already been developed with the data, and you can find their results here.

Given this dataset, use pandas to create a baseline model (i.e. a model that just predicts the most common class) and answer the following questions:





In [56]:
pet_df = pd.read_clipboard()
pet_df.head()

actual model1 model2 model3 model4
0    cat    cat    dog    cat    dog
1    dog    dog    cat    cat    dog
2    dog    cat    cat    cat    dog
3    dog    dog    dog    cat    dog
4    cat    cat    cat    dog    dog

In [57]:
pet_df.actual.value_counts()

actual
dog    3254
cat    1746
Name: count, dtype: int64

In [58]:
#baseline
(pet_df.actual == 'dog').mean()

0.6508

In [59]:
3253/(3254+1746)

0.6506

dog is my positive class

a. In terms of accuracy, how do the various models compare to the baseline model? 

model1 has the highest accuracy with 81%

Are any of the models better than the baseline?

model1(81%) and model4(74%) are higher than the baseline model of 65%.

accuracy = 

TP + TN /

TP + TN + FP + FN


In [75]:
pets1 = pd.crosstab(pet_df.actual, pet_df.model1)
pets2 = pd.crosstab(pet_df.actual, pet_df.model2)
pets3 = pd.crosstab(pet_df.actual, pet_df.model3)
pets4 = pd.crosstab(pet_df.actual, pet_df.model4)
pets1
pets2
pets3
pets4

model4  cat   dog
actual           
cat     603  1143
dog     144  3110

In [62]:
pet1_tp = pets1.iloc[0,0] #rows, columns
pet1_fn = pets1.iloc[0,1]
pet1_fp = pets1.iloc[1,0]
pet1_tn = pets1.iloc[1,1]

In [63]:
pet2_tp = pets2.iloc[0,0] #rows, columns
pet2_fn = pets2.iloc[0,1]
pet2_fp = pets2.iloc[1,0]
pet2_tn = pets2.iloc[1,1]

In [64]:
pet3_tp = pets3.iloc[0,0] #rows, columns
pet3_fn = pets3.iloc[0,1]
pet3_fp = pets3.iloc[1,0]
pet3_tn = pets3.iloc[1,1]

In [65]:
pet4_tp = pets4.iloc[0,0] #rows, columns
pet4_fn = pets4.iloc[0,1]
pet4_fp = pets4.iloc[1,0]
pet4_tn = pets4.iloc[1,1]

In [61]:
(pet_df.actual == pet_df.model1).mean()

0.8074

In [66]:
(pet1_tp + pet1_tn) / (pet1_tp + pet1_tn + pet1_fp + pet1_fn)

0.8074

In [67]:
(pet_df.actual == pet_df.model2).mean()

0.6304

In [68]:
(pet2_tp + pet2_tn) / (pet2_tp + pet2_tn + pet2_fp + pet2_fn)

0.6304

In [69]:
(pet_df.actual == pet_df.model3).mean()

0.5096

In [70]:
(pet3_tp + pet3_tn) / (pet3_tp + pet3_tn + pet3_fp + pet3_fn)

0.5096

In [71]:
(pet_df.actual == pet_df.model4).mean()

0.7426

In [72]:
(pet4_tp + pet4_tn) / (pet4_tp + pet4_tn + pet4_fp + pet4_fn)

0.7426

b. Suppose you are working on a team that solely deals with dog pictures. Which of these models would you recommend?

In [77]:
print(pets1)
print('')
print(pets2)
print('')
print(pets3)
print('')
print(pets4)

model1   cat   dog
actual            
cat     1423   323
dog      640  2614

model2   cat   dog
actual            
cat     1555   191
dog     1657  1597

model3   cat   dog
actual            
cat      893   853
dog     1599  1655

model4  cat   dog
actual           
cat     603  1143
dog     144  3110


>precision to minimize false positives to avoid getting cat pictures

In [86]:
dog_pos_predictions1 = pet_df [pet_df.model1 == 'dog']

In [87]:
(dog_pos_predictions.actual == dog_pos_predictions1.model1).mean()

0.8900238338440586

In [88]:
dog_pos_predictions2 = pet_df [pet_df.model2 == 'dog']

In [90]:
(dog_pos_predictions2.actual == dog_pos_predictions2.model2).mean()

0.8931767337807607

In [91]:
dog_pos_predictions3 = pet_df [pet_df.model3 == 'dog']

In [92]:
(dog_pos_predictions3.actual == dog_pos_predictions3.model3).mean()

0.6598883572567783

In [93]:
dog_pos_predictions4 = pet_df [pet_df.model4 == 'dog']

In [94]:
(dog_pos_predictions4.actual == dog_pos_predictions4.model4).mean()

0.7312485304490948

>model 1 or model 2 would be best for predicting dog pictures

c. Suppose you are working on a team that solely deals with cat pictures. Which of these models would you recommend?

>precision to minimize false positives to avoid dog pictures

In [95]:
for col in pet_df.columns[1:]:

    pos_predictions = pet_df [pet_df[col] == 'cat']
    prec = (pos_predictions.actual == pos_predictions[col]).mean()
    
    print(f'for {col} the precision is {round(prec,2)}')

for model1 the precision is 0.69
for model2 the precision is 0.48
for model3 the precision is 0.36
for model4 the precision is 0.81


>model4 would be the best fit based on precision

5. Follow the links below to read the documentation about each function, then apply those functions to the data from the previous problem.

sklearn.metrics.accuracy_score
sklearn.metrics.precision_score
sklearn.metrics.recall_score
sklearn.metrics.classification_report

In [96]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, classification_report

In [97]:
accuracy_score(pet_df.actual, pet_df.model1)

0.8074

In [99]:
precision_score(pet_df.actual, pet_df.model1, pos_label='dog')

0.8900238338440586

In [98]:
precision_score(pet_df.actual, pet_df.model1, pos_label='cat')

0.6897721764420747

In [100]:
recall_score(pet_df.actual, pet_df.model1, pos_label='cat')

0.8150057273768614

In [101]:
recall_score(pet_df.actual, pet_df.model1, pos_label='dog')

0.803318992009834

In [102]:
classification_report(pet_df.actual, pet_df.model1)

'              precision    recall  f1-score   support\n\n         cat       0.69      0.82      0.75      1746\n         dog       0.89      0.80      0.84      3254\n\n    accuracy                           0.81      5000\n   macro avg       0.79      0.81      0.80      5000\nweighted avg       0.82      0.81      0.81      5000\n'

In [103]:
print(classification_report(pet_df.actual, pet_df.model1))

              precision    recall  f1-score   support

         cat       0.69      0.82      0.75      1746
         dog       0.89      0.80      0.84      3254

    accuracy                           0.81      5000
   macro avg       0.79      0.81      0.80      5000
weighted avg       0.82      0.81      0.81      5000

